sklearn_numba_dpex/common/topk.py

```diff
@@ -0,0 +1,692 @@
+            k_in_subset,
+            radix_position,
+            mask_for_desired_value,
+            desired_masked_value,
+            # OUT
+            terminate,
+        )
+
+        if sequential_processing_on_different_device:
+            desired_masked_value = desired_masked_value.to_device(device)
+
+        if int(terminate[0]) == 1:
+            break
+
+        if sequential_processing_on_different_device:
+            radix_position = radix_position.to_device(device)
+            mask_for_desired_value = mask_for_desired_value.to_device(device)
+            desired_masked_value = desired_masked_value.to_device(device)
+
+        initialize_privatized_counts(privatized_counts)
+
+    if sequential_processing_on_different_device:
+        k_in_subset = k_in_subset.to_device(device)
+
+    # reinterpret the threshold back to a dtype item
+    threshold = dpt.usm_ndarray(
+        shape=desired_masked_value.shape, dtype=dtype, buffer=desired_masked_value
+    )
+
+    return k_in_subset, threshold, work_group_size, dtype, device
+
+
+# The Top-K implemented in this file uses radix sorting. Radix sorting consists in
+# using lexicographical order on bits to sort the items. It assumes that the
+# lexicographical order on the bits of the items that are being sorted matches their
+# natural ordering.
+
+# However, this is not true for sets that mix positive and negative items, given that
+# the dtypes that are supported set the first bit to 1 to encode the negative sign, and
+# 0 to encode the positive sign, in such a way that negative floats are lesser than
+# positive floats using the natural order while the opposite is true with
+# lexicograpbical order on bits.
+
+# Still, it is possible to enable radix sorting for dtypes that don't follow the
+# lexicographical order, providing a bijection can be found with an unsigned type, such
+# as the lexicographical order on the unsigned dtype matches the natural order on the
+# input dtype. The target dtype must also support bitwise operations. Thus, float32 is
+# mapped to uint32, float64 is mapped to uint64,...
+
+# In practice such bijections can be found for all dtypes and are cheap to compute. The
+# code that follows defines bijections to unsigned integers for float32 and float64
+# floats, with the following bitwise transformations:
+#    - reinterpret the float as an unsigned integer (i.e the same buffer used to encode
+#      a float is used to encode an unsigned integer. Said differently, first transform
+#      the float to the integer such as the bits used to encode the integer are the
+#      same than the bits used to encode the float)
+#    - if the float item is positive, set the first bit of the unsigned integer to 1
+#    - if the float item is negative, flip all of the bits of the unsigned integer
+#      (i.e set to 1 if bit is 0, else 0)
+#
+# See also:
+# https://stackoverflow.com/questions/4640906/radix-sort-sorting-a-float-data/\
+#     4641059#4641059
```

Comment:
Should all of those rationales comes at the top of this file?

Reply:
Moved at the top of the file